### ISSUE TITLE
Multiple RemoteWebDrivers per jBrowserDriver process in a Selenium Grid2?

### ISSUE THREAD

**Issue by smcnamara3** [MachinePublishers/jBrowserDriver#94]
It looks like the RMI stuff within jBrowserDriver is being designed to support multiple clients simultaneously through the Selenium API. However I am using a Grid2 configuration and there doesn't seem to be any way to do that.

Specifically, it seems like this is happening:

1. Selenium client does `new RemoteWebDriver()`
2. Selenium Grid2 Hub server matches the request caps up with one of the available Selenium Grid2 Nodes based on the caps parameters and proxies the request data through the hub
3. Selenium Grid2 Node locally calls `new JBrowserDriver()` (eventually) similar to how one would do it locally

Problem is, each `JBrowserDriver`, as I understand it, is a system _process_ - which is pretty heavy and wasteful if what you can do works fine in-process with shared state and a bunch of threads.

Would be nice if the Grid2 Node could use an existing JBrowserDriver until the JBD is fully allocated (given the number of concurrent requests it advertises itself as being able to service) and give the client an instance of RemoteWebDriver that eventually talks to the _same_ JBD process.

That way, the client calling `new RemoteWebDriver()` twice, and hitting the same Selenium Grid2 Node, with a JBD behind it that supports, say, 50 concurrent sessions, would only involve a total of four JVMs:

1. The Selenium Client JVM
2. The Selenium Grid2 Hub Server JVM
3. The Selenium Grid2 Node JVM
4. The JBrowserDriver JVM (servicing each WebDriver with a different thread)

vs. the current model:

1. The Selenium Client JVM
2. The Selenium Grid2 Hub Server JVM
3. The Selenium Grid2 Node JVM
4. The JBrowserDriver JVM #1 (servicing one WebDriver)
5. The JBrowserDriver JVM #2 (servicing the other WebDriver)

**Comment by hollingsworthd** [comment]
True but it's not more wasteful than other remote web drivers (such as chrome, firefox, etc) since those are communicating with a browser in a separate process. In the case of this project, JBD *is* the browser in the separate process.

One thing that could be more efficient in this context is that WebDriver.quit() kills the JRE process, but it could actually keep it alive so that the next new instance can just re-use that process. Already this is what JBrowserDriver.reset(..) is doing.

As far as making this all share one process and using separate threads instead, the complexity is that Java's WebView doesn't support that out of the box (if you want truly separate sessions). So it involves some work there, and also a number of changes to the project itself. I'm not planning on implementing that. What I might eventually do though is allow a user to run a max of one browser instance and not have it spawn a new process (everything runs in the callers' process). This might be useful for people in constrained environments where spawning a new process is not possible... but that's not really addressing your concern.

So in summary, I don't think there are any more OS processes with JBrowserDriver in this setup than with any other remote web driver... it's just that those processes are JREs instead of Chrome/Firefox.

**Comment by hollingsworthd** [comment]
Also I think I should clarify, any time someone is calling `new JBrowserDriver(..)` they are generally going to want a separate thread for that and subsequent calls to that instance if they're interested in having multiple browser instances. So you'd want N threads for N calls to the JBrowserDriver constructor.

Some people have misinterpreted that within a single instance there are someone N number of browsers but that's not the case. The Selenium API doesn't really allow for that.

Further, if you're familiar with how the RemoteWebDriver class works internally, you may be confused as to what's going one. JBrowserDriver, while extending RemoteWebDriver, doesn't actually use any of its functionality. Most RemoteWebDrivers will start a browser that implements the selenium wire protocol then RemoteWebDriver passes commands to it. JBrowserDriver overrides all of that functionality to replace it all with RMI calls to spawned process of JBrowserDriver itself.
